src/protocol/boolean/dumb_bitwise_add_constant.rs

```diff
@@ -0,0 +1,192 @@
+/// which is the bit-decomposition of `a + B`.
+///
+/// Note that the index notation of the inputs is `0..l-1`, whereas the output
+/// index notation is `0..l`. This means that the output of this protocol will be
+/// "`l+1`"-bit long bitwise secret shares, where `l = |[a]_B|`.
+///
+/// Really simple logic. Just follows the way you do addition in grade school
+/// Starting from the least significant digit add up the digits, carrying when required.
+/// We can skip a few multiplications because one input has known values for each bit.
+///
+/// The tricky part is in computing the "carries".
+/// The logic is slightly different for the least significant bit. For that, we just look
+/// at the value of the least significant bit of the constant. If it is a zero, then there
+/// is no way there is any carry, and the answer is just zero. If it is a one, then the carry
+/// will be a one, only if the least significant bit of a `[a_0]_p` is a secret-sharing of one.
+/// So it's just equal to `[a_0]_p`.
+///
+/// For all subsequent carries, once again, it depends on the value of the constant bit at that place:
+/// - when the constant has a bit value of zero - there is only a carry if BOTH of the previous carry
+/// AND the secret bit are shares of one. We can get this result by just multiplying the two.
+///
+/// When the constant has a bit value of one - there is a carry if EITHER the previous carry
+/// OR the secret bit are shares of one. We can get this result using OR (x OR y = x + y - x*y).
+/// So in either case, we need to multiply the previous carry with the next bit of the secret input.
+///
+/// As such, there are a total of l-2 multiplications (one less than the bit-length of the input).
+/// Sometimes, a multiplication can be skipped, because we know, a prioi that the result must be zero.
+pub async fn bitwise_add_constant<F, C, S>(
+    ctx: C,
+    record_id: RecordId,
+    a: &[S],
+    b: u128,
+) -> Result<Vec<S>, Error>
+where
+    F: Field,
+    C: Context<F, Share = S>,
+    S: ArithmeticSecretSharing<F>,
+{
+    let mut output = Vec::with_capacity(a.len() + 1);
+
+    let mut last_carry_known_to_be_zero = (b & 1) == 0;
+    let mut last_carry = if last_carry_known_to_be_zero {
+        S::ZERO
+    } else {
+        a[0].clone()
+    };
+    let result_bit = if last_carry_known_to_be_zero {
+        a[0].clone()
+    } else {
+        ctx.share_of_one() - &a[0]
+    };
+    output.push(result_bit);
+
+    for (bit_index, bit) in a.iter().enumerate().skip(1) {
+        let mult_result = if last_carry_known_to_be_zero {
+            // TODO: this makes me sad
+            let _ = ctx
+                .narrow(&BitOpStep::from(bit_index))
+                .multiply(record_id, &S::ZERO, &S::ZERO) // this is stupid
+                .await?;
+
+            S::ZERO
+        } else {
+            ctx.narrow(&BitOpStep::from(bit_index))
+                .multiply(record_id, &last_carry, bit)
+                .await?
+        };
```

Comment:
```suggestion
        let mult_result = if last_carry_known_to_be_zero {
            async move { Ok::<_, Error>(S::ZERO) }.left_future()
        } else {
            ctx.narrow(&BitOpStep::from(bit_index))
                .multiply(record_id, &last_carry, bit)
                .right_future()
        }
        .await?;
```

Reply:
Unfortunately, this stupid multiply is necessary until we work something else out.  We need to ensure that there is a contiguous series of records for each step and with the random revealed value, this optimization leaves gaps at the low bits any time there is a contiguous series of zero bits.

Ultimately, we'll have to build something to track how many preceding records had these gaps so that we can adjust the record ID for these multiplies down (that is, for bit B, record R, how many in records between 0 and R-1 had B-1 or more zero-valued bits in the revealed value, then subtract that from R to get the functional record ID).

Reply:
Oh I see. So this fails in other tests like oneshot/ipa? I ran the unit tests in this file and it checked out.

Reply:
Yes IPA test cries due to missing processing of records :(